File: scripts/polar_line.py
import math
import cv2
import numpy as np


class PolarLine:
    def __init__(self, line=None):
        self.theta = 0
        self.rho = 0
        
        if line is not None:
            # Zamiana odcinka na parametry prostej w postaci polarnej.
            dx = line[2] - line[0]
            dy = line[3] - line[1]
            
            self.theta = math.atan2(dy, dx) + math.pi / 2
            self.rho = line[0] * math.cos(self.theta) + line[1] * math.sin(self.theta)
            
            # Ujednolicenie znaku rho, aby podobne linie mialy porownywalna reprezentacje.
            if self.rho < 0:
                self.rho = -self.rho
                self.theta += math.pi
            
            # Normalizacja kata do zakresu <0, 2pi).
            self.theta = self.theta % (2 * math.pi)
            while self.theta < 0:
                self.theta += 2 * math.pi

    def __eq__(self, other):
        if not isinstance(other, PolarLine):
            return False
        
        theta_error = abs(self.theta - other.theta)
        rho_error = abs(self.rho - other.rho)
        
        # Dwie linie traktujemy jako rowne, jesli maja zblizony kat i polozenie.
        if theta_error < 0.1 and rho_error < 200:
            return True
        else:
            return False

    def __lt__(self, other):
        return self.rho < other.rho

    def is_horizontal(self):
        # Kierunek normalnej pozwala rozroznic, czy linia jest blizej poziomej czy pionowej.
        return abs(math.sin(self.theta)) > abs(math.cos(self.theta))

    def draw(self, image, color):
        a = math.cos(self.theta)
        b = math.sin(self.theta)
        x0 = a * self.rho
        y0 = b * self.rho
        
        # Rysowanie bardzo dlugiego odcinka, aby pokazac cala prosta na obrazie.
        point1 = (int(x0 - 20000 * (-b)), int(y0 - 20000 * a))
        point2 = (int(x0 + 20000 * (-b)), int(y0 + 20000 * a))
        cv2.line(image, point1, point2, color, 10)


def collect_document_lines(line_segments):
    horizontal_lines = []
    vertical_lines = []
    
    for line_segment in line_segments:
        polar_line = PolarLine(line_segment)
        
        if polar_line.is_horizontal():
            # Dodaj tylko nowe reprezentanty podobnych linii poziomych.
            if polar_line not in horizontal_lines:
                horizontal_lines.append(polar_line)
        else:
            # Dodaj tylko nowe reprezentanty podobnych linii pionowych.
            if polar_line not in vertical_lines:
                vertical_lines.append(polar_line)
    
    # Sortowanie po rho ulatwia pozniejszy wybor skrajnych krawedzi dokumentu.
    horizontal_lines.sort()
    vertical_lines.sort()
    
    return horizontal_lines, vertical_lines

File: src/polar_line.cpp
#include "polar_line.h"

#include <algorithm>
#include <cmath>

using namespace cv;

PolarLine::PolarLine(Vec4i line)
{
    double dx = line[2] - line[0];
    double dy = line[3] - line[1];

    // Zamiana odcinka na parametry prostej w postaci polarnej.
    theta = atan2(dy, dx) + M_PI_2;
    rho = line[0] * cos(theta) + line[1] * sin(theta);

    // Ujednolicenie znaku rho, aby podobne linie mialy porownywalna reprezentacje.
    if (rho < 0)
    {
        rho = -rho;
        theta += CV_PI;
    }

    // Normalizacja kata do zakresu <0, 2pi).
    theta = fmod(theta, 2 * CV_PI);
    while (theta < 0)
    {
        theta += 2 * CV_PI;
    }
}

bool PolarLine::operator==(const PolarLine& other) const
{
    double thetaError = abs(theta - other.theta);
    double rhoError = abs(rho - other.rho);

    // Dwie linie traktujemy jako rowne, jesli maja zblizony kat i polozenie.
    if (thetaError < 0.1 && rhoError < 200)
    {
        return true;
    }
    else
    {
        return false;
    }
}

bool PolarLine::operator<(const PolarLine& other) const
{
    return rho < other.rho;
}

bool PolarLine::isHorizontal() const
{
    // Kierunek normalnej pozwala rozroznic, czy linia jest blizej poziomej czy pionowej.
    return abs(sin(theta)) > abs(cos(theta));
}

void PolarLine::draw(Mat image, Scalar color) const
{
    double a = cos(theta);
    double b = sin(theta);
    double x0 = a * rho;
    double y0 = b * rho;

    // Rysowanie bardzo dlugiego odcinka, aby pokazac cala prosta na obrazie.
    Point point1(x0 - 20000 * (-b), y0 - 20000 * a);
    Point point2(x0 + 20000 * (-b), y0 + 20000 * a);
    line(image, point1, point2, color, 10);
}

std::pair<std::vector<PolarLine>, std::vector<PolarLine>> collectDocumentLines(const std::vector<Vec4i>& lineSegments)
{
    std::vector<PolarLine> horizontalLines;
    std::vector<PolarLine> verticalLines;
    for (const auto& lineSegment : lineSegments)
    {
        PolarLine polarLine(lineSegment);

        if (polarLine.isHorizontal())
        {
            // Dodaj tylko nowe reprezentanty podobnych linii poziomych.
            auto existingLine = std::find(horizontalLines.begin(), horizontalLines.end(), polarLine);
            if (existingLine == horizontalLines.end())
            {
                horizontalLines.push_back(polarLine);
            }
        }
        else
        {
            // Dodaj tylko nowe reprezentanty podobnych linii pionowych.
            auto existingLine = std::find(verticalLines.begin(), verticalLines.end(), polarLine);
            if (existingLine == verticalLines.end())
            {
                verticalLines.push_back(polarLine);
            }
        }
    }

    // Sortowanie po rho ulatwia pozniejszy wybor skrajnych krawedzi dokumentu.
    std::sort(horizontalLines.begin(), horizontalLines.end());
    std::sort(verticalLines.begin(), verticalLines.end());

    return {horizontalLines, verticalLines};
}